File: advance/src/eu/advance/logistics/flow/model/AdvanceTypeVariable.java
/*
 * Copyright 2010-2012 The Advance EU 7th Framework project consortium
 *
 * This file is part of Advance.
 *
 * Advance is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Advance is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with Advance.  If not, see
 * <http://www.gnu.org/licenses/>.
 *
 */

package eu.advance.logistics.flow.model;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.List;

import com.google.common.collect.Lists;

import eu.advance.logistics.xml.typesystem.XElement;

/**
 * The type variable definition of an Advance block.
 * It allows
 * @author karnokd, 2011.07.01.
 */
public class AdvanceTypeVariable {
	/** The type parameter name. */
	public String name;
	/** The upper bounds of this type parameter, e.g., T super SomeObject1 &amp; SomeObject2. */
	public final List<AdvanceType> bounds = Lists.newArrayList();
	/** Indicator if the bounds are representing the upper bound. */
	public boolean isUpperBound;
	/** The documentation explaining this type variable. */
	public URI documentation;
	/**
	 * Load the contents from an XML element with a schema of <code>block-description.xsd</code>.
	 * @param root the root element
	 */
	public void load(XElement root) {
		name = root.get("name");
		for (XElement ub : root.childrenWithName("upper-bound")) {
			isUpperBound = true;
			AdvanceType t = new AdvanceType();
			t.load(ub);
			bounds.add(t);
		}
		for (XElement lb : root.childrenWithName("lower-bound")) {
			if (isUpperBound) {
				throw new IllegalArgumentException("Type variable has multiple bound types! " + root);
			}
			AdvanceType t = new AdvanceType();
			t.load(lb);
			bounds.add(t);
		}
		String u = root.get("documentation");
		if (u != null) {
			try {
				documentation = new URI(u);
			} catch (URISyntaxException ex) {
				throw new RuntimeException(ex);
			}
		}
	}
}

File: advance-api-cs/AdvanceAPIClient/Classes/Model/AdvanceTypeVariable.cs
/*
 * Copyright 2010-2013 The Advance EU 7th Framework project consortium
 *
 * This file is part of Advance.
 *
 * Advance is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Advance is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with Advance.  If not, see
 * <http://www.gnu.org/licenses/>.
 *
 */
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Xml;

using AdvanceAPIClient.Core;
using AdvanceAPIClient.Interfaces;

namespace AdvanceAPIClient.Classes.Model
{
    /// <summary>
    /// The type variable definition of an Advance block.
    /// </summary>
    public class AdvanceTypeVariable : XmlReadWrite
    {
        /// <summary>
        /// type parameter name
        /// </summary>
        public string Name;
        /// <summary>
        /// Upper bounds of this type parameter, e.g., T super SomeObject1 &amp; SomeObject2. 
        /// </summary>
        public List<AdvanceType> Bounds = new List<AdvanceType>();
        /// <summary>
        /// Indicator if the bounds are representing the upper bound.
        /// </summary>
        public bool IsUpperBound;
        /// <summary>
        /// Documentation explaining this type variable.
        /// </summary>
        public Uri Documentation;

        public AdvanceTypeVariable() : base() {}

        /// <summary>
        /// Create a deep copy of this type variable
        /// </summary>
        /// <returns>new type variable</returns>
        public AdvanceTypeVariable Copy()
        {
            AdvanceTypeVariable result = new AdvanceTypeVariable();
            result.Name = this.Name;
            result.IsUpperBound = this.IsUpperBound;
            result.Documentation = this.Documentation;
            foreach (AdvanceType b in this.Bounds)
            {
                result.Bounds.Add(b.Copy());
            }
            return result;
        }

        public override string ToString()
        {
            StringBuilder b = new StringBuilder();

            b.Append(this.Name);
            if (this.Bounds.Count > 0)
            {
                if (this.IsUpperBound)
                    b.Append(" super ");
                else
                    b.Append(" extends ");
                int i = 0;
                foreach (AdvanceType bound in this.Bounds)
                {
                    if (i > 0) b.Append(" & ");
                    b.Append(bound);
                }
            }
            return b.ToString();
        }

        /// <summary>
        /// Fill object data from xml node
        /// </summary>
        /// <param name="source"></param>
        protected override void LoadFromXmlNode(XmlNode source)
        {
            this.Name = GetAttribute(source, "name", null);
            this.IsUpperBound = false;
            foreach (XmlNode node in GetChildren(source, "upper-bound"))
            {
                this.IsUpperBound = true;
                this.Bounds.Add(CreateFromXml<AdvanceType>(node));
            }
            foreach (XmlNode node in GetChildren(source, "lower-bound"))
            {
                if (this.IsUpperBound)
                    this.ThrowException("Type variable has multiple bound types! ");
                this.Bounds.Add(CreateFromXml<AdvanceType>(node));
            }
            this.Documentation = GetUriAttribute(source, "documentation");
        }

        /// <summary>
        /// Fills  object's xml
        /// </summary>
        /// <param name="node">Empty node for the object with name "tagname"</param>
        protected override void FillXmlElement(XmlElement node)
        {
            AddAttribute(node, "name", this.Name);
            string boundName = this.IsUpperBound ? "upper-bound" : "lower-bound";
            foreach (AdvanceType t in this.Bounds)
                t.AddToXML(boundName, node);
            if (this.Documentation != null)
                AddAttribute(node, "documentation", this.Documentation);
        }
    }
}


